### PR TITLE
[ROCm] reduce tolerance for triangular solve with well_conditioned set to True

### PR DESCRIPTION
**Pull request by alugorey**
Current test case causes an edge case tensor input that causes a single generated tensor to fail the tolerance assertion on ROCm only and only for float32. We have reviewed the logic with our libraries team and have discovered the discrepancy is due to a difference in order of operations on AMD GPUs. They came back with "working as intended" and found no perceivable bug. Interestingly, if we change the values in ks, ns, or bs, the test passes on ROCm. These particular sizes in this particular order generates a single problematic input that causes the assertion to fail the tolerance check by ~0.07. Again, this is not a bug, just differences in implementation. This PR loosens the tolerance for ROCm only.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

File: test/test_linalg.py [hunk 1]
```diff
@@ -4043,11 +4043,9 @@ def _test_linalg_solve_triangular(self, A, B, upper, left, uni):
         self.assertEqual(X, out)
 
     @dtypes(*floating_and_complex_types())
-    @precisionOverride({torch.float32: 1e-1, torch.complex64: 1e-1,
+    @precisionOverride({torch.float32: 2e-1 if TEST_WITH_ROCM else 1e-1, torch.complex64: 1e-1,
                         torch.float64: 1e-8, torch.complex128: 1e-8})
```
Comment (nikitaved) [inline test/test_linalg.py:4047]: Hmm, I wonder whether we could just modify inputs to be better conditioned? `1e-1` seems very low to me...

**Review by jeffdaily — APPROVED** [review]

File: test/test_linalg.py (hunk 1 above)
Comment (nikitaved) [reply test/test_linalg.py:4047]: @alugorey , `_test_linalg_solve_triangular` has a `well_conditioned` argument which is `False` by default. Could you please try this one instead and reduce the tolerance level. `1e-1` is just too low to be able to tell the function is doing the right thing.

Comment (alugorey) [reply test/test_linalg.py:4047]: @nikitaved setting well_conditioned to true allows the test to pass. pushing up that change now

Comment (alugorey) [reply test/test_linalg.py]: @nikitaved I was able to bring it down to 1e-4 and still have it pass. 1e-4 fails. Would you like me to peg it there?

Comment (nikitaved) [reply test/test_linalg.py]: `1e-4` seems fine, it is both for `float` and `cfloat`, right?

Comment (alugorey) [reply test/test_linalg.py]: 1e-4 for float32, i was able to push complex64 to 1e-8 and still pass

Comment (nikitaved) [reply test/test_linalg.py]: Let's just remove explicit tolerance for cfloat then!

**Review by nikitaved — APPROVED** [review]
Nice! Thank you, now it looks much better!

File: test/test_linalg.py (hunk 1 above)
Comment (alugorey) [reply test/test_linalg.py]: Done! removed for both complex types, both pass!

Comment (alugorey) [reply test/test_linalg.py]: @nikitaved uh oh! looks like 1e-4 is too tight for CPU test cases which are now failing

**Review by lezcano — APPROVED** [review]

**Comment by alugorey** [comment]
@nikitaved Hello! I know you approved this, but we are now seeing tolerance failures for f32 on CPU tests only. Wondering if you had a recommendation on how you wanted to address this? We could increase tolerance slightly (1e-3?) or we can make these tolerances be GPU or ROCm only? Thoughts?

**Comment by nikitaved** [comment]
@alugorey , I am fine with that for now, it is still much better than `1e-1`. We can investigate the CPU path in a follow-up...

File: test/test_linalg.py (hunk 1 above)
Comment (nikitaved) [reply test/test_linalg.py]: Could you please just leave a comment on the current choose and that it is dictated by the CPU path?

**Comment by eqy** [comment]
Does this break `linux-bionic-cuda11.8-py3.10-gcc7-debug / test (default, 4, 5, linux.4xlarge.nvidia.gpu)`?
e.g., https://github.com/pytorch/pytorch/actions/runs/5429187622/jobs/9874360641
CC @huydhn

**Comment by huydhn** [comment]
> Does this break `linux-bionic-cuda11.8-py3.10-gcc7-debug / test (default, 4, 5, linux.4xlarge.nvidia.gpu)`? e.g., https://github.com/pytorch/pytorch/actions/runs/5429187622/jobs/9874360641 CC @huydhn

You're right, I think it breaks that test.  it's a surprise that only [debug build](https://github.com/pytorch/pytorch/blob/main/setup.py#L5C1-L6) fails.  I'm not exactly sure what the solution would be, so I guess it would be cleaner to revert and reland this change

**Comment by alugorey** [comment]
@nikitaved 
Looks like this ended up failing other tolerance tests. Thoughts on where to go from here?

**Comment by huydhn** [comment]
The error looks unusual IMO as it fails only on CUDA test built with debug https://github.com/pytorch/pytorch/actions/runs/5429187622/jobs/9874360641.  It was a CPU test that failed

So I think this is what we can do to unblock:

* Reopen this PR, rebase, and rerun the test.  I will add `ciflow/periodic` to make sure that the test run with PyTorch debug build
* If it still fails, let's gate the change to only ROCm because it looks like an unrelated issue like what you say

**Comment by nikitaved** [comment]
Hmm, the non-ROCM-related things need to get investigated, these are very strange...

**Pull request closed without merging**